### DISCUSSION TITLE
Option to Change background and font color

### DISCUSSION THREAD

**Discussion by Nick222** [zim-desktop-wiki/zim-desktop-wiki#1802] [Q&A]
How user can change background and font color Zim 0.70 ?

**Comment by jaap-karssenberg** [comment]
See https://zim-wiki.org/manual/Help/Config_Files.html

Probably you need to use the CSS file in this case

**Comment by Nick222** [comment]
It is not simple, as I see...

Can You create options for it ?

**Reply by hunnybuny** [reply]
It is not located where most users say it is:
/home/[user]/.var/app/org.zim_wiki.Zim/config/zim/style.conf

Replace [user] with your own login name. Happy editing!

**Comment by jaap-karssenberg** [comment]
Anything is possible, but my personal time is limited. If anyone wants to
work on a good patch I'm more than willing to merge it.

**Comment by Nick222** [comment]
But why in ~/.config/zim/style.conf

[TextView]
background=#cccccc

do not work ?

**Comment by jaap-karssenberg** [comment]
The reason that that doesn't work is that the "Styles" config only controls gtk.TextTag style. The background color and font color are controlled by the Gtk theme, so it is a different type of configuration.

**Comment by vivasteve** [comment]
If you are using zim on linux you can change the gtk theme - this is now easier than before v.0.70 since Jaap has so spent so much time refactoring zim to use gtk-3 (much appreciated!)

gtk3 settings are in ~/.config/gtk-3.0
you can change the theme with gtk-theme-name
available themes are in /usr/share/themes/
check in each theme dir that a gtk-3.0 dir is present - then those themes are avail for gtk-3.0

not sure how this might work in windows ...

**Comment by Nick222** [comment]
**vivasteve**
But I want to change background colour only in Zim - not in all software...

**Comment by sojusnik** [comment]
@Nick222 You can try the [TextView](https://omacronides.com/projets/zim-plugin-textview-prefs/) plugin.

**Comment by vivasteve** [comment]
> But I want to change background colour only in Zim - not in all software...

I see your point - for me it is ok because zim is the only gtk app I use!

**Comment by lucvtzel** [comment]
> **vivasteve**
> But I want to change background colour only in Zim - not in all software...

You can do it by creating a `~/.config/gtk-3.0/gtk.css` file like this (tested with *GTK3.18-Ubuntu16.04*):

```
#zim-pageview {
    background-color: #2E3436;
    color : #FFFFFF;
    /* font  : "Ubuntu Mono"; NO !: the right place for the font setting is in '~/.config/zim/style.conf' */
}

#zim-pageview:selected {
  background-color: #6699cc;
}

#zim-pageindex {
    color : #FFFFFF;
    background-color : #2E3436;
}

#zim-tags-tagcloud {
    background-color : #2E3436;
}

#zim-tags-tagcloud GtkLabel {
    color: #EDD400;
}
```

**Comment by Nick222** [comment]
Very good! :)
But where can I get all used by Zim style`s name (zim-pageview and so on)?

**Comment by lucvtzel** [comment]
**jaap-karssenberg** [told us](https://github.com/zim-desktop-wiki/zim-desktop-wiki/issues/755#issuecomment-479489905) to see in the manual (see section `Gtk configuration` in the 'Config Files' page or at [https://zim-wiki.org/manual/Help/Config_Files.html](https://zim-wiki.org/manual/Help/Config_Files.html)) :

> To identify these in the code, search for use of ``widget_set_css(...)`` and ``widget.set_name(...)``. The name used will be accessible from the CSS config as an "id".

For instance, if we examine the code in the `pageview.py` file we will find [this line](https://github.com/zim-desktop-wiki/zim-desktop-wiki/blob/8d9c573f5b46aa7576c3ea712c7e2d04405ec369/zim/gui/pageview.py#L3518) in the constructor: 

	self.set_name('zim-pageview')

We can also see all the ID names from the GTK Inspector. These links can be helpful:

+ [How do I open GTK Inspector?](https://askubuntu.com/a/748152)
+ [GTK+ CSS Overview](https://developer.gnome.org/gtk3/stable/chap-css-overview.html)
+ [Using the CSS file format](https://developer.gnome.org/gtk3/stable/gtk-migrating-GtkStyleContext-css.html)

Below a way to explore the code and find the use of ``widget_set_css(...)`` and ``widget.set_name(...)``:
```
$ grep -rn -e "\(\.set_name(\s*['\"]zim\)" && grep -rn -e "\(\widget_set_css(.*,\s\+['\"]\)"
```
```
zim/gui/pageview.py:3524:		self.set_name('zim-pageview')
zim/gui/widgets.py:2072:		self.set_name('zim-sidepane')
zim/plugins/pathbar.py:477:		self.set_name('zim-pathbar')
zim/plugins/pageindex/__init__.py:361:		self.set_name('zim-pageindex')
zim/plugins/tags.py:347:		self.set_name('zim-tags-tagcloud')
zim/gui/insertedobjects.py:58:		widget_set_css(self._vbox, 'zim-inserted-object', 'border: 1px solid #ccc')
zim/gui/insertedobjects.py:70:		widget_set_css(widget, 'zim-inserted-object-head', 'border-bottom: 1px solid #ccc')
zim/gui/widgets.py:750:	widget_set_css(button, 'zim-statusbar-button',	'''
zim/plugins/pathbar.py:189:		widget_set_css(child, 'zim-pathbar-path-button', 'padding: 2px 6px 2px 6px;')
zim/plugins/pathbar.py:467:		widget_set_css(self, 'zim-pathbar-scroll-button', 'padding: 2px 2px 2px 2px;')
zim/plugins/tableofcontents.py:484:		widget_set_css(self, 'zim-toc-widget', 'border: 1px solid @fg_color')
zim/plugins/tableofcontents.py:485:		widget_set_css(self.head, 'zim-toc-head', 'border-bottom: 1px solid @fg_color')
zim/plugins/distractionfree.py:130:		#widget_set_css(self._maxwidth, 'zim-distractionfree-bin',
```

**Comment by Foxblade** [comment]
> @Nick222 You can try the [TextView](https://omacronides.com/projets/zim-plugin-textview-prefs/) plugin.

I installed the plugin to the correct folder but don't seem to see it appear in the plugins list in order to be enabled. Any ideas? 

[Plugins window after placing the files in the plugins folder](https://i.imgur.com/YZTjBaf.png)

**Comment by Nick222** [comment]
>  Any ideas?

There are two versions of this plugin - for Zim 0.68 and for Zim 0.71 (downloads from same page)

**Comment by asdfqy** [comment]
> You can do it by creating a `~/.config/gtk-3.0/gtk.css` file like this (tested with _GTK3.18-Ubuntu16.04_):
> 
> ```
> #zim-pageview {
>     background-color: #2E3436;
>     color : #FFFFFF;
>     font  : "Ubuntu Mono";
> }
> 
> #zim-pageview:selected {
>   background-color: #6699cc;
> }
> 
> #zim-pageindex {
>     color : #FFFFFF;
>     font : "Ubuntu Mono";
>     background-color : #2E3436;
> }
> 
> #zim-tags-tagcloud {
>     background-color : #2E3436;
> }
> 
> #zim-tags-tagcloud GtkLabel {
>     color: #EDD400;
> }
> ```

somehow for me the syntax had to be a little bit different
the hint to use *selection* is from https://developer.gnome.org/gtk3/stable/GtkTextView.html
the usage of the stars is from trial-and-error .. maybe someone can explain on this 
```
#zim-pageview* {
    background-color: #2F4F4F;
    color : #8FBABE;
}
#zim-pageview selection {
  background-color: #6699cc;
  color: #FFFFFF;
}
#zim-sidepane*  {
    background-color: #2F4F4F;
    color : #8FBABE;
}
#zim-sidepane*:selected {
  background-color: #6699cc;
}
```

**Comment by lucvtzel** [comment]
This [commit 547f2750fba...](https://github.com/zim-desktop-wiki/zim-desktop-wiki/commit/547f2750fba09d582607d16648d8fb18a309ba42?diff=unified) from [arigit](https://github.com/zim-desktop-wiki/zim-desktop-wiki/commits?author=arigit) is also interesting. 
I would add that to launch the GTK+ Inspector, we can use the Control-Shift-I or Control-Shift-D keyboard shortcuts as it it is [mentioned](https://developer.gnome.org/gtk3/stable/gtk-running.html#interactive-debugging) in the GTK documentation.

**Comment by Nick222** [comment]
Move to Discussion, pls...

**Comment by beezyphile** [comment]
I tried to add some margins and padding but couldn't do it. I'm really new to editing css, is it possible to do. My eyes hurt when the text is leaning too close to the borders.

**Reply by introt** [reply]
I'm afraid I can't answer your question, but I can suggest trying out the Distraction Free Editing plugin

**Reply by introt** [reply]
There's also `justify=CENTER`, which centers all text, moving it away from the border, but that's likely not what you're going after either

**Reply by bruvellu** [reply]
You can adjust the text's border width using the `Textview Preferences` plugin. Give it a try @beezyphile

**Reply by introt** [reply]
You can find the plugin via the links in #2576

**Comment by opuslux** [comment]
I managed to set the background-color (yellow) for everything I need in the gtk.css but not for the Attachements- and Backlings tabs in the sidepane _when they are empty_. When no attachments or backlinks are present they stay white:

![image](https://github.com/zim-desktop-wiki/zim-desktop-wiki/assets/6863949/4167ed10-ef58-46a6-b65b-961239f77df7)

Any hints on this would be very appreciated.

**Comment by brucewestfall** [comment]
Glad to see this here!  It is an excellent resource to look back on when needed.  I appreciate the variety on this forum.  I have asked questions on others and either been ignored or told to "Google it".  

Thanks to everyone for being willing to help out and patience with us who are still learning.
